### ISSUE TITLE
Only evaluate `with_columns` on returned rows on `LazyFrame`

### ISSUE THREAD

**Issue by qiemem** [pola-rs/polars#9527]
### Problem description

Currently, any involvement of `with_columns` on a `LazyFrame` in which the added columns are returned appears to require a full evaluation of those columns, even for rows that are not returned. For example, I have a 4.6G parquet dataset, scanned via `scan_parquet` (though the results are the same with, for instance, `scan_csv` or `scan_pyarrow_dataset`):

```python
In [16]: %time df.select("[run number]").head(5).collect()
CPU times: user 2.96 ms, sys: 0 ns, total: 2.96 ms
Wall time: 11.4 ms
Out[16]: 
shape: (5, 1)
┌──────────────┐
│ [run number] │
│ ---          │
│ i64          │
╞══════════════╡
│ 6            │
│ 3            │
│ 4            │
│ 1            │
│ 19           │
└──────────────┘

In [17]: %time df.with_columns(foo=5).select("[run number]", "foo").head(5).collect()
CPU times: user 1.07 s, sys: 227 ms, total: 1.3 s
Wall time: 1.3 s
Out[17]: 
shape: (5, 2)
┌──────────────┬─────┐
│ [run number] ┆ foo │
│ ---          ┆ --- │
│ i64          ┆ i32 │
╞══════════════╪═════╡
│ 6            ┆ 5   │
│ 3            ┆ 5   │
│ 4            ┆ 5   │
│ 1            ┆ 5   │
│ 19           ┆ 5   │
└──────────────┴─────┘
```

Evaluation of `foo` for the returned rows should not require the evaluation of `foo` for every row in the dataset, but the difference in time indicates that unnecessary evaluation is happening.  It would be awesome if polars could detect when expressions in `with_columns` can wait to be evaluated until after filtering, slicing, etc operations (effectively moving the `with_columns` to after the `head` in this case).

Now, of course, there are expressions which require operating on complete columns (aggregations, window functions, etc). But there are a large class of simple, one-to-one operations that do not.

The specific motivation for this is that I often use `.with_columns` directly after a scan to do things like converting columns to more convenient forms, and it would be great to not have to worry about that later in the analysis. Current use is casting hive partition values read in via `scan_pyarrow_dataset` from strings to, eg, numeric values. Right now, I have to repeatedly do that after other operations, while it would be much more convenient to put it right after the scan and not have to worry about it anymore.

Thanks as always for creating this amazing library! More recently, `scan_pyarrow_dataset` has been a total game changer for my workflow. Really awesome to be able to use partitioned datasets.

**Comment by deanm0000** [comment]
Have you tried using `.profile` instead of `collect`? 

`head` deliberately does all the work and is excluded from query optimizations. `fetch`, seemingly, would do what you want, at least in the example.

**Comment by qiemem** [comment]
Fixed by  #12742
